### PR TITLE
Cachable S3 URLs

### PR DESCRIPTION
**Pull request by joedevivo**
### Never tell me the odds!

Every time we create a signed S3 URL, it expires in X minutes. Since the expiration time is included in the URL's query string, as seconds since the epoch (1/1/70)

The existing logic to compute the expiry time

```erlang
expiration_time(TimeToLive) ->
    Epoch = calendar:datetime_to_gregorian_seconds({{1970, 1, 1}, {0, 0, 0}}),
    Now = calendar:datetime_to_gregorian_seconds(erlang:universaltime()),
    (Now - Epoch) + TimeToLive.
```

So, you can't cache these. You just can't. Unless it was requested in the same second because of the call to `erlang:universaltime()`. So what to do?

Well, what if we set an interval size. In the following example, I'll use 15 minute intervals, and a TimeToLive of 1 hour.

We divide the day into intervals, so every URL generated in that interval is the same.

```
 +--A--a--aaa---+---------B----+--C---ccccc---+-----Dddd----d+
1:00           1:15           1:30           1:45           2:00
```

Look at the URL `A` genrated at 1:03, until 1:15 they'll all be set to expire at 2:15. These are the lowec case `a` URLs, which just means they're the same, but they're not "new".

Once it's 1:15, we're in a new inteval, but we don't try getting a URL until 1:25. We set that one to expire at 2:30 (1hr + the remaining time in the interval). 

Nobody asks for a URL until after 1:30, so the URL `B` is only used once and is never asked for again. Oh well. We played the odds and lost this time. It's not the end of the world.

At 1:33, the URL `C` is generated and this interval gets used alot, so it's good that we have this feature.

You get the idea. Over the course of the day we will only ever generate 96 unique expiration times, as opposed to a new expiration time for every URL requested.

Now, 15m may not be the optimal window size. If we went with 60m then we'd only generate 24 unique URLs per day. That's why we've made it configurable.

Take the existing API for generating a URL:

```erlang
-spec s3_url(atom(), string(), string(), integer(),
             proplists:proplist(), config()) -> binary().
s3_url(Method, BucketName, Key, Lifetime, RawHeaders, Config)
```

`Lifetime` is an `integer()`. The number of second that this thing should expire in. This would have been `3600`, the number of seconds in an hour. If you still do it this way, it will work just as it did before.

We've added a tuple `{Lifetime, IntervalSize}` version of the argument. 

```erlang
-spec s3_url(atom(), string(), string(), integer() | {integer(), integer()},
             proplists:proplist(), config()) -> binary().
s3_url(Method, BucketName, Key, Lifetime, RawHeaders, Config)
```

In the above example with a 15m interval, we'd have passed in `{3600, 900}` which would essentially make the lifetime = 3600 + (some number between 0 and 900).

### CODE REVIEW

File: src/mini_s3.erl [hunk 1]
```diff
@@ -459,7 +483,7 @@ s3_url(Method, BucketName, Key, Lifetime, RawHeaders,
     Path = lists:flatten([$/, BucketName, $/ , Key]),
     CanonicalizedResource = ms3_http:url_encode_loose(Path),
 
-    {_StringToSign, Signature} = make_signed_url_authorization(SecretKey, Method,
+    {_StringToSign, Signature} = mini_s3:make_signed_url_authorization(SecretKey, Method,
```
Comment (sdelano) [inline src/mini_s3.erl]: Just curious, what's the reasoning for making this call fully qualified with `mini_s3:...` now?

**Comment by sdelano** [comment]
:+1: Are there any branches of `oc_erchef` with this integrated in yet?

File: src/mini_s3.erl (hunk 1 above)
Comment (joedevivo) [inline src/mini_s3.erl]: good :eyes:, it came down to meck not mocking calls that weren't fully qualified

**Comment by joedevivo** [comment]
No branches yet. I figured this was contained enough. Will get to oc_erchef next.

**Comment by ryancragun** [comment]
:100:

**Pull request merged**